sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcUtil.java

```diff
@@ -1,73 +1,77 @@
-                    break;
-                case INT32: // four-byte signed integer.
-                    preparedStatement.setInt(index + 1, element.getInt32(index));
-                    break;
-                case INT64: // eight-byte signed integer.
-                    preparedStatement.setLong(index + 1, element.getInt64(index));
-                    break;
-                case DECIMAL: // Arbitrary-precision decimal number
-                    preparedStatement.setBigDecimal(index + 1, element.getDecimal(index));
-                    break;
-                case FLOAT:
-                    preparedStatement.setFloat(index + 1, element.getFloat(index));
-                    break;
-                case DOUBLE:
-                    preparedStatement.setDouble(index + 1, element.getDouble(index));
-                    break;
-                case STRING: // String.
-                    preparedStatement.setString(index + 1, element.getString(index));
-                    break;
-                case DATETIME: // Date and time.
-                    preparedStatement.setTimestamp(index + 1, new Timestamp(element.getDateTime(index).getMillis()));
-                    break;
-                case BOOLEAN: // Boolean.
-                    preparedStatement.setBoolean(index + 1, element.getBoolean(index));
-                    break;
-                case BYTES: // Byte array.
-                    preparedStatement.setBytes(index + 1, element.getBytes(index));
-                    break;
                 case ARRAY:
                     preparedStatement.setArray(index + 1,
                             preparedStatement.getConnection().createArrayOf(fieldType.getCollectionElementType().getTypeName().name(), element.getArray(index).toArray()));
                     break;
-                case MAP:
-                    throw new RuntimeException("Map in schema is not supported while writing. Please provide statement and preparedStatementSetter");
-                case ROW: // The field is itself a nested row.
-                    throw new RuntimeException("Row in schema is not supported while writing. Please provide statement and preparedStatementSetter");
                 case LOGICAL_TYPE: {
                     String logicalTypeName = fieldType.getLogicalType().getIdentifier();
                     JDBCType jdbcType = JDBCType.valueOf(logicalTypeName);
                     switch (jdbcType) {
                         case DATE:
                             preparedStatement.setDate(index + 1, new Date(element.getDateTime(index).getMillis()));
+                            break;
                         case TIME:
                             preparedStatement.setTime(index + 1, new Time(element.getDateTime(index).getMillis()));
+                            break;
                         case TIMESTAMP_WITH_TIMEZONE:
                             preparedStatement.setTimestamp(index + 1, new Timestamp(element.getDateTime(index).getMillis()));
```

Comment:
So the idea is to determine the set of prepared statement setters once to avoid having to execute a switch statement for each row that you want to insert. This refactoring hasn't fully achieved that effect yet. Please take a careful look at `BeamRowMapper` implementation again to see what I am talking about.